modules/images/image-loading-optimization/detect.js

```diff
@@ -0,0 +1,279 @@
+					}
+					resolve();
+				},
+				{
+					root: null, // To watch for intersection relative to the device's viewport.
+					threshold: 0.0, // As soon as even one pixel is visible.
+				}
+			);
+
+			for ( const element of breadcrumbedElementsMap.keys() ) {
+				if ( ! adminBar || ! adminBar.contains( element ) ) {
+					intersectionObserver.observe( element );
+				}
+			}
+		} );
+
+		// Stop observing as soon as the page scrolls since we only want initial-viewport elements.
+		win.addEventListener( 'scroll', disconnectIntersectionObserver, {
+			once: true,
+			passive: true,
+		} );
+	}
+
+	// TODO: Use a local copy of web-vitals.
+	const { onLCP } = await import(
+		// eslint-disable-next-line import/no-unresolved
+		'https://unpkg.com/web-vitals@3/dist/web-vitals.js?module'
+	);
+
+	/** @type {LCPMetric[]} */
+	const lcpMetricCandidates = [];
+
+	// Obtain at least one LCP candidate. More may be reported before the page finishes loading.
+	await new Promise( ( resolve ) => {
+		onLCP(
+			( metric ) => {
+				lcpMetricCandidates.push( metric );
+				resolve();
+			},
+			{
+				// This avoids needing to click to finalize LCP candidate. While this is helpful for testing, it also
+				// ensures that we always get an LCP candidate reported. Otherwise, the callback may never fire if the
+				// user never does a click or keydown, per <https://github.com/GoogleChrome/web-vitals/blob/07f6f96/src/onLCP.ts#L99-L107>.
+				reportAllChanges: true,
+			}
+		);
+	} );
+
+	// Wait until the resources on the page have fully loaded.
+	await new Promise( ( resolve ) => {
+		if ( doc.readyState === 'complete' ) {
+			resolve();
+		} else {
+			win.addEventListener( 'load', resolve, { once: true } );
+		}
+	} );
+
+	// Stop observing.
+	disconnectIntersectionObserver();
+	if ( isDebug ) {
+		log( 'Detection is stopping.' );
+	}
+
+	/** @type {PageMetrics} */
+	const pageMetrics = {
+		viewport: {
+			width: win.innerWidth,
+			height: win.innerHeight,
+		},
+		elements: [],
+	};
+
+	const lcpMetric = lcpMetricCandidates.at( -1 );
+
+	for ( const elementIntersection of elementIntersections ) {
+		const breadcrumbs = breadcrumbedElementsMap.get(
+			elementIntersection.target
+		);
+		if ( ! breadcrumbs ) {
+			if ( isDebug ) {
+				warn( 'Unable to look up breadcrumbs for element' );
+			}
+			continue;
+		}
+
+		const isLCP =
+			elementIntersection.target === lcpMetric?.entries[ 0 ]?.element;
+
+		/** @type {ElementMetrics} */
+		const elementMetrics = {
+			isLCP,
+			isLCPCandidate: !! lcpMetricCandidates.find(
+				( lcpMetricCandidate ) =>
+					lcpMetricCandidate.entries[ 0 ]?.element ===
+					elementIntersection.target
+			),
+			breadcrumbs,
+			intersectionRatio: elementIntersection.intersectionRatio,
+			intersectionRect: elementIntersection.intersectionRect,
+			boundingClientRect: elementIntersection.boundingClientRect,
+		};
+
+		pageMetrics.elements.push( elementMetrics );
+	}
+
+	// TODO: Send data to server.
+	log( pageMetrics );
```

Comment:
This is to be done in another PR. See https://github.com/WordPress/performance/pull/878.